Source/Particles/Collision/BinaryCollision/PhotonPhoton/SinglePhotonPhotonCollisionEvent.H

```diff
@@ -0,0 +1,183 @@
+/* Copyright 2021 Neil Zaim
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#ifndef SINGLE_PHOTON_PHOTON_COLLISION_EVENT_H_
+#define SINGLE_PHOTON_PHOTON_COLLISION_EVENT_H_
+
+#include "LinearBreitWheelerCrossSection.H"
+
+#include "Particles/Collision/BinaryCollision/BinaryCollisionUtils.H"
+#include "Utils/WarpXConst.H"
+
+#include <AMReX_Algorithm.H>
+#include <AMReX_Random.H>
+#include <AMReX_REAL.H>
+
+#include <cmath>
+
+
+/**
+ * \brief This function computes whether the collision between two photons results in a
+ * pair-producing (electron and positron) event, using the algorithm described in
+ * Higginson et al., Journal of Computational Physics 388, 439-453 (2019).
+ * If the electron-positron pair is produced, the mask is set to true for that given pair
+ * of macrophotons and the weight of the produced particles is stored in p_pair_reaction_weight.
+ *
+ * Note that the word <<pair>> sometimes refers to the electron-positron products of the collision,
+ * sometimes to the macrophotons that are paired to collide in a cell according to the algorithm.
+ * It should be clear from the context.
+ *
+ * @tparam index_type type of the index argument
+ * @param[in] u1x,u1y,u1z momenta of the first colliding particle
+ * @param[in] u2x,u2y,u2z momenta of the second colliding particle
+ * @param[in] w1,w2 effective weight of the colliding particles
+ * @param[in] dt is the time step length between two collision calls.
+ * @param[in] dV is the volume of the corresponding cell.
+ * @param[in] pair_index is the index of the colliding pair
+ * @param[out] p_mask is a mask that will be set to true if pair (electron-positron) production
+ * occurs for that pair of macrophotons
+ * @param[out] p_pair_reaction_weight stores the weight of the product particles
+ * @param[in] event_multiplier factor used to increase the number of pair-production events by
+ * decreasing the weight of the produced particles
+ * @param[in] multiplier_ratio factor used to take into account unsampled pairs (i.e. the fact
+ * that a particle only collides with one or few particles of the other species)
+ * @param[in] probability_threshold probability threshold above which we decrease the event
+ * multiplier
+ * @param[in] probability_target_value if the probability threshold is exceeded, this is used
+ * to determine by how much the event multiplier is reduced
+ * @param[in] engine the random engine.
+ */
+template <typename index_type>
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void SinglePhotonPhotonCollisionEvent (const amrex::ParticleReal& u1x, const amrex::ParticleReal& u1y,
+                               const amrex::ParticleReal& u1z, const amrex::ParticleReal& u2x,
+                               const amrex::ParticleReal& u2y, const amrex::ParticleReal& u2z,
+                               amrex::ParticleReal w1, amrex::ParticleReal w2,
```

Comment:
I would suggest using something like:
```suggestion
                               const amrex::ParticleReal& w1, const amrex::ParticleReal& w2,
```